Consider disabled armaments invalid

reaperrr opened the pull request:
Fixes #13855.

Possibly a candidate for stable, since it's an easy and obvious fix.

reaperrr committed:
Consider disabled armaments invalid

reaperrr commented:
@MustaphaTR Can you confirm that this fixes #13855?

MustaphaTR commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -209,6 +209,9 @@ public bool HasAnyValidWeapons(Target t, bool checkForCenterTargetingWeapons = f
 			foreach (var armament in Armaments)
 			{
+				if (armament.IsTraitDisabled)
+					continue;
```
Not tested yet, but isn't this should be `!armament.IsTraitDisabled` or `return false;`. We want it to work if armament is enabled, not disabled.

reaperrr replied:
`continue` means if this `armament` is disabled, it will skip to and test the next armament and so on, until it either a) comes across one that is enabled and valid and therefore returns `true`, or b) all are disabled or invalid and the method returns `false`.

MustaphaTR replied:
Ok, you are right. After testing you are right this fixes the problem.

In fact i tried to fix that myself too and put the check below as 	

    if (checkIsValid && !armament.IsTraitDisabled && armament.Weapon.IsValidAgainst(t, self.World, self))

so i think that's why i got confused.

pchote replied:
IMO @MustaphaTR's version is more readable, since it makes it clear that this part of the same validity check, and not something distinct.  Would you mind adjusting this?

reaperrr replied:
Done.

MustaphaTR reviewed (approved):
👍

reaperrr commented:
Updated.

pchote commented:
At the risk of prolonging this PR further... `CombatDebugOverlay`, `AttackGarrisoned`, and `FlyAttack` look like the need similar fixes.  Find uses of `AttackBase.Armaments` to confirm.

It may make more sense to move the check into the definition of `getArmaments` instead?

reaperrr committed:
Properly account for disabled Armaments in various places

These places didn't care if an Armament was disabled, which could lead to unexpected behavior.

reaperrr commented:
> It may make more sense to move the check into the definition of getArmaments instead?

`getArmaments` currently runs only once at actor creation.
I don't see a simple, performant way to change that, so I've gone for fixing it at the places that use AB.Armaments.

pchote commented:
> `getArmaments` currently runs only once at actor creation.

For this I was thinking of the `.ToArray().Where(Exts.TraitIsEnabled)` enumerable caching that we do in several other places.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2c90ec329ea40eba7b83d5a62baaf2eb4ba5c57b

3038f15f00f2da3734efffb9bc2cf9c6ed00fa32
56547360860b054f44719d514a2296a18027670a
